crates/starknet-types-rpc/src/custom_serde.rs

```diff
@@ -0,0 +1,137 @@
+use core::marker::PhantomData;
+
+/// A trait for types that should be serialized or deserialized as hexadecimal strings.
+pub trait NumAsHex<'de>: Sized {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer;
+
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>;
+}
+
+impl<'de> NumAsHex<'de> for u64 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        const HEX_DIGITS: &[u8] = b"0123456789abcdef";
+        const MAX_NUMBER_SIZE: usize = u64::MAX.ilog(16) as usize + 1;
+
+        // This can be very much optimized using unsafe code.
+        // Let's benchmark it first to ensure that it's actually worth it.
+
+        let mut buffer = [0u8; MAX_NUMBER_SIZE + 2]; // + 2 to account for 0x
+        let mut cursor = buffer.iter_mut().rev();
+        let mut n = *self;
+        while n != 0 {
+            *cursor.next().unwrap() = HEX_DIGITS[(n % 16) as usize];
+            n /= 16;
+        }
+        *cursor.next().unwrap() = b'x';
+        *cursor.next().unwrap() = b'0';
+
+        let remaining = cursor.len();
+
+        // SAFETY:
+        //  We only wrote ASCII characters to the buffer, ensuring that it is only composed
+        //  of valid UTF-8 code points.
+        let s = core::str::from_utf8(&buffer[remaining..]).unwrap();
```

Comment:
why bother doing all that instead of just 
```rs
format!("{self:x"}");
```

Reply:
To avoid to things:

1. Having to gothrough the Rust `fmt::Arguments` type which uses dynamic dispatch (meaning it cannot be optimized by `rustc`)
2. Havint to allocate a whole `String` every time a number has to be serialized (which is very common).